File: go/patterns-use-cases/src/dataupload/service/datauploadservice.go
package main

import (
	"context"
	restate "github.com/restatedev/sdk-go"
	"github.com/restatedev/sdk-go/server"
	"log/slog"
	"os"
)

type DataUploadService struct{}

func (DataUploadService) Run(ctx restate.WorkflowContext) (string, error) {
	url, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
		return createS3Bucket(), nil
	})
	if err != nil {
		return "", err
	}

	_, err = restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
		return restate.Void{}, uploadData(url)
	})
	if err != nil {
		return "", err
	}

	err = restate.Promise[string](ctx, "url").Resolve(url)
	if err != nil {
		return "", err
	}

	return url, nil
}

func (DataUploadService) ResultAsEmail(ctx restate.WorkflowSharedContext, email string) error {
	slog.Info("Slow upload: client requested to be notified via email.")
	url, err := restate.Promise[string](ctx, "url").Result()
	if err != nil {
		return err
	}

	_, err = restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
		return restate.Void{}, sendEmail(email, url)
	})
	return err
}

func main() {
	server := server.NewRestate().
		Bind(restate.Reflect(DataUploadService{}))

	if err := server.Start(context.Background(), ":9080"); err != nil {
		slog.Error("application exited unexpectedly", "err", err.Error())
		os.Exit(1)
	}
}

File: typescript/patterns-use-cases/src/syncasync/data_upload_service.ts
import * as restate from "@restatedev/restate-sdk";
import {
  WorkflowContext,
  WorkflowSharedContext,
} from "@restatedev/restate-sdk";
import { createS3Bucket, sendEmail, uploadData } from "./utils";

const dataUploadService = restate.workflow({
  name: "dataUploader",
  handlers: {
    run: async (ctx: WorkflowContext) => {
      const url = await ctx.run(() => createS3Bucket());
      await ctx.run(() => uploadData(url));

      await ctx.promise<URL>("url").resolve(url);
      return url;
    },

    resultAsEmail: async (
      ctx: WorkflowSharedContext,
      req: { email: string }
    ) => {
      console.info("Slow upload: client requested to be notified via email.");
      const url = await ctx.promise<URL>("url");
      await ctx.run(() => sendEmail(req.email, url));
    },
  },
});

export type DataUploadService = typeof dataUploadService;

restate.endpoint().bind(dataUploadService).listen(9080);
